File: solana/ts/src/matchingEngine/state/PreparedSlowOrder.ts
import { BN } from "@coral-xyz/anchor";
import { PublicKey } from "@solana/web3.js";

export class PreparedSlowOrder {
    bump: number;
    preparedBy: PublicKey;
    fastVaaHash: Array<number>;
    sourceChain: number;
    baseFee: BN;

    constructor(
        bump: number,
        preparedBy: PublicKey,
        fastVaaHash: Array<number>,
        sourceChain: number,
        baseFee: BN
    ) {
        this.bump = bump;
        this.preparedBy = preparedBy;
        this.fastVaaHash = fastVaaHash;
        this.sourceChain = sourceChain;
        this.baseFee = baseFee;
    }

    static address(programId: PublicKey, payer: PublicKey, fastVaaHash: Array<number>) {
        return PublicKey.findProgramAddressSync(
            [Buffer.from("prepared"), payer.toBuffer(), Buffer.from(fastVaaHash)],
            programId
        )[0];
    }
}

File: solana/programs/matching-engine/src/state/prepared_slow_order.rs
use anchor_lang::prelude::*;

#[account]
#[derive(Debug, InitSpace)]
pub struct PreparedSlowOrder {
    pub bump: u8,
    pub prepared_by: Pubkey,
    pub fast_vaa_hash: [u8; 32],

    pub source_chain: u16,
    pub base_fee: u64,
}

impl PreparedSlowOrder {
    pub const SEED_PREFIX: &'static [u8] = b"prepared";
}
